docs/source/diagnostics.rst

```diff
@@ -0,0 +1,86 @@
+Diagnostics
+===========
+
+Profiling parallel code can be tricky. ``dask.diagnostics`` provides
+functionality to aid in profiling and inspecting dask graph execution.
+
+
+Scheduler Callbacks
+-------------------
+
+Schedulers based on ``dask.async.get_async`` (currently
+``dask.async.get_sync``, ``dask.threaded.get``, and
+``dask.multiprocessing.get``) accept two callbacks, allowing for inspection of
+dask execution. The callbacks are:
+
+1. ``start_callback(self, key, dask, state)``
+    Callback ran every time a new task is started. Receives the key of the
+    task to be run, the dask, and the scheduler state. The final callback
+    receives ``None`` instead of a key, as no new tasks were added at that
+    tick.
+2. ``end_callback(self, key, dask, state, id)``
+    Callback ran every time a task is finished. Receives the key of the
+    task to be run, the dask, the scheduler state, and the id of the worker
+    that ran the task. The final callback receives ``None`` none for both key
+    worker id, as no new tasks were finished at that tick.
+
+
+ Callbacks for common use cases are provided in ``dask.diagnostics``.
+
+
+Profiler
+--------
+
+The ``Profiler`` class profiles dask execution at the task level. To use,
```

Comment:
Might be nice to show connection from the previous section.

The ``Profiler`` class builds on the scheduler callbacks to profile ...